instance/models/openedx_instance.py

```diff
@@ -213,17 +214,34 @@ def get_load_balancer_configuration(self, triggered_by_instance=False):
             )
         return backend_map, backend_conf
 
-    def set_active_vm_dns_records(self):
+    def set_active_vm_dns_records(self, deactivate_appserver=False):
         """
         Set DNS A records for all active app servers.
         """
         self.logger.info("Setting DNS records for active app servers...")
-        for i, appserver in enumerate(self.get_active_appservers(), 1):
-            ip_addr = appserver.server.public_ip
-            if ip_addr:
+        with cache.lock('set_active_vm_dns_records_' + str(self.ref.instance_id)):
+            for i, appserver in enumerate(self.get_active_appservers(), 1):
+                ip_addr = appserver.server.public_ip
+
+                if ip_addr:
+                    domain = "vm{index}.{base_domain}".format(index=i, base_domain=self.internal_lms_domain)
+                    gandi.api.set_dns_record(domain, type="A", value=ip_addr)
+
+            if deactivate_appserver:
+                active_appservers = self.get_active_appservers()
+                unused_dns_index = active_appservers.count() + 1
+                domain = "vm{index}.{base_domain}".format(index=unused_dns_index, base_domain=self.internal_lms_domain)
+                gandi.api.remove_dns_record(domain, type="A")
+
+    def clean_up_appserver_dns_records(self):
+        """
+        Removes the DNS records for the app servers.
+        """
+        self.logger.info("Cleaning up DNS records for app servers...")
+        with cache.lock('clean_vm_dns_records_' + str(self.ref.instance_id)):
```

Comment:
@pkulkark, nit: better to use `str.format` instead of `+`.

Reply:
@pkulkark, the locking is useful here only when using the same lock as in the `set_active_vm_dns_records` method. There is no risk in two concurrent archive operations running at the same time as that will not leave any dangling DNS records. However, if a new appserver is activated at the same time as the instance is archived (I know it is highly unlikely, but still possible), there is a chance that one DNS record maybe left undeleted.

Since the lock is used in 2 places, it might be better to name it something like `appserver_dns_record_update` instead.

Reply:
Sure. I updated it.